nixos/influxdb2: add provisioning and nixos tests

oddlama opened the pull request:
## Description of changes

This PR adds provisioning options and nixos tests to influxdb2. This allows skipping the initial setup page and automatically creates/deletes the following entities:
- organizations
- buckets
- users
- api tokens (auths)

Originally initial setup and entity provisioning was a single PR which has been split into two to reduce scope and discuss details of the implementation. This is the second part, which needs to be rebased the first part once merged:
- #249211 

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

oddlama commented:
In summary, @NickCao was asking whether the setup script can be simplified (and thus was split to this separate PR):
> Overall I like the idea of declarative setup, but the implementation looks overly complicated. Can we first get the part that can be done with influx setup in and figure out better ways to do the rest?

I'm currently not seeing how this could be made simpler, but I am open for suggestions.

oddlama commented:
Rebased on master.

oddlama commented:
Restructured the provisioning script. All scripts are now defined on the respective option for increased consistency and readability. Also added the option to provision the secret value of api tokens.

NickCao commented on nixos/modules/services/databases/influxdb2.nix:
```diff
@@ -3,18 +3,133 @@
   configFile = format.generate "config.json" cfg.settings;
+
+  # A helper utility to allow provisioning tokens with deterministic secrets
+  tokenManipulator = pkgs.buildGoModule rec {
```
This should be a separate package not to be inlined in the module.

oddlama replied:
True, I should probably load a json file instead of baking the configuration in.

NickCao commented on nixos/modules/services/databases/influxdb2.nix:
```diff
@@ -3,18 +3,133 @@
+
+    export INFLUX_TOKEN=$(< "$CREDENTIALS_DIRECTORY/admin-token")
+
+    ${concatMapStrings (x: x._script) cfg.provision.deleteApiTokens}
```
Instead of `delete`, naming it `absent` would be more *declarative*.

NickCao reviewed (commented):
And I thinks that a better API might look like:
```nix
provisioning.resources = {
  organization = {
    acme-corp = {
      present = true; # the default
      some-attribute = "bar";
      bucket = {
        example = {
          present = false; # indicates removal
        };
      };
    };
  };
}
```

NickCao commented:
And as there are various influxdb modules for ansible (https://docs.ansible.com/ansible/latest/collections/community/general/influxdb_database_module.html), we can generate a ansible-playbook to carry out the actual operation.

oddlama commented:
> And I thinks that a better API might look like: [...]

Fully agree, I'll refactor this.

> And as there are various influxdb modules for ansible (https://docs.ansible.com/ansible/latest/collections/community/general/influxdb_database_module.html), we can generate a ansible-playbook to carry out the actual operation.

I fear that is a module for influx v1 not v2. I've also searched for alternatives, but I cannot find any that support provisioning remotes, replications or authentication tokens.

oddlama commented on nixos/modules/services/databases/influxdb2.nix:
```diff
@@ -76,17 +191,783 @@ in
+                    influx bucket list --json --org ${escapeShellArg config.org} --name ${escapeShellArg config.name} 2>/dev/null \
+                      | jq -r ".[0].id"
+                  ); then
+                    influx bucket delete --id "$id" >/dev/null
```
Interesting, the cli is a little inconsistent at times unfortunately. Would you prefer to deviate from the usual schema in this instance?

NickCao replied:
Aligning with the usual schema is preferable if does save us some duplicate code.

NickCao commented:
Another thought: we may use https://github.com/influxdata/influxdb-client-python of other sdks instead of the cli, which would be easier/more readable than bash

oddlama commented:
Summary of the last update:

- I've switched to using an external python script that is utilizing influxdb-client and applies the desired state which is passed via json.
- For now, I've removed provisioning of remotes and replications from the scope of this PR, because it will require more work (the library has the capability but doesn't contain all required classes to do so easily). Maybe for a future PR.
- Added extensive testing
- Shortly mention this change in the release notes
- Rebased on master

NickCao commented on nixos/modules/services/databases/influxdb2.nix:
```diff
@@ -3,18 +3,125 @@
+    set -euo pipefail
+    if test -e "$STATE_DIRECTORY/.needs_restart"; then
+      rm -f "$STATE_DIRECTORY/.needs_restart"
+      systemctl restart influxdb2
```
It does not feel like correct running sytemctl in systemd units.

oddlama replied:
I agree, but I have no other solution to the problem. Suggestions welcome.

NickCao commented on nixos/modules/services/databases/influxdb2.nix:
```diff
@@ -57,13 +164,9 @@ in
-            '';
+            type = types.ints.unsigned;
+            default = 0;
+            description = "The duration in seconds for which the bucket will retain data (0 is infinite).";
```
Actually we should add `lib.mdDoc` to all these descriptions.

NickCao commented on nixos/modules/services/databases/influxdb2.nix:
```diff
@@ -111,58 +448,38 @@ in
           touch "$STATE_DIRECTORY/.first_startup"
+        else
+          # Manipulate provisioned api tokens if necessary
+          ${getExe pkgs.influxdb2-token-manipulator} "$STATE_DIRECTORY/influxd.bolt" ${tokenMappings}
```
Do we have other options than directly modifying the database?

oddlama replied:
No, unfortunately this is the only way. There is no API allowing for declarative tokens except for the initial admin token (where we already use it).

NickCao commented on nixos/modules/services/databases/influxdb2.nix:
```diff
@@ -76,17 +179,251 @@ in
+            }
+          '';
+          default = {};
+          type = types.attrsOf (types.submodule (organizationSubmod: let
```
Also we may move these submodule declarations to the let/in block above, making the module less cluttered.

oddlama commented on nixos/modules/services/databases/influxdb2.nix:
```diff
@@ -111,58 +448,38 @@ in
+        ExecStartPost = mkIf cfg.provision.enable (
+          [provisioningScript] ++
+          # Only the restarter runs with elevated privileges
+          optional anyAuthDefined "+${restarterScript}"
```
I've already tried realizing it using systemd alone, but I've not found a working configuration. The main issue is that RestartForceExitStatus only works when the main process exits with that status, and not pre or post scripts:

> Takes a list of exit status definitions that, when returned by the main service process [...]

NickCao replied:
Grepped for systemctl in `nixos/modules` and found a vast number of invocations, guess that's ok. Just many of them hardcode the path to systemctl as `/run/current-system/systemd/bin/systemctl`

oddlama replied:
And the influxdb service absorbs all signals except for SIGTERM, which I don't want to use for restarting because systemd also uses it to signal the service to exit. So if the service should be stopped, it would first be restarted and then killed after a timeout with SIGKILL. That's why I finally opted to do `systemctl restart influxdb2`, because it just was the only solution I found that doesn't mess with normal behavior.

oddlama commented:
Update:

- Use lib.mdDoc for descriptions
- Move organizations submodule declaration to top let block
- Use /run/current-system/systemd/bin/systemctl instead of systemctl

NickCao commented:
And be sure to split the changes into individual commits. (I did not check the correctness of the assertions e.t.c. but I'm sure you can do a better job than me in that regard)

oddlama committed:
influxdb2-token-manipulator: init at 1.0.0

oddlama committed:
influxdb2-provision: init at 1.0.0

oddlama committed:
nixos/influxdb2: add org, bucket, users and auth provisioning

oddlama commented:
Split in to 3 commits (1 for each package, one for the module update)

NickCao commented:
@ofborg test influxdb2

NickCao reviewed (commented):
Evaluation with provision disabled currently fails with 
```
       Failed assertions:
       - influxdb2: provision.organizations.someorg.auths.sometoken: The `operator` and `allAccess` options are mutually exclusive with each other and the granular permission settings.
  ```

My bad, forgot to also modify the specialisation part.

NickCao reviewed (approved):
Tested with nixos tests, also tested evaluation with provision disabled.
